return error when create volume with existing name

allencloud opened the pull request:
fixes #16068

Currently, `docker volume create --name volumeA` will not return error when volumeA already exists in daemon side, and it will return some info about the existing volume.
In addition, `docker volume create --name volumeA --label a=b` will not add label for volumeA which seems a little improper.

In the code implemented related to this Part. I think it is not so reasonable that volume name can be the same while with different volume drivers. Since it will also affect `docker volume inspect volumeA` if volumeA has two.

This PR addressed the above issue.

**\- What I did**
1. make creating volume with existing name return an error
2. make volume name unique among all volume drivers.

**\- How I did it**

**\- How to verify it**

**\- Description for the changelog**

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

**\- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: allencloud allen.sun@daocloud.io

allencloud committed:
return error when create volume with existing name

Signed-off-by: allencloud <allen.sun@daocloud.io>

tonistiigi commented:
@allencloud CI is failing

ssuvasanth commented:
@allencloud This affects the UX of docker cli.
If you run `docker volume create --name volumeA` and then use it in your container `docker run -v volumeA:/mountpoint -it ubuntu`, you always get `errNameConflict`.

This needs a design review to decide if we should return the volume on finding a match from the same volume driver or a conflict.

I wish to know your thoughts @cpuguy83 @AkihiroSuda @tonistiigi

allencloud commented:
@ssuvasanth What you said is totally correct. Thanks a lot. 😄 
Actually, it is too late for me to working on this issue. And I will keep on it soon.

allencloud commented:
Thanks for your review. @tonistiigi @ssuvasanth 
I tried to address some situation on UX of cli. Docs are also modified, and api change as well. 
And I added more information in the pr description. PTAL
Please tell me if I missed something.

ssuvasanth commented:
@allencloud I think `docker run -v <existing volume name>:/mountpoint image` would still throw errNameConflict. Could you please confirm?

allencloud commented:
@ssuvasanth 
Thanks for your review.
I try to address your provided issue and did a test. Here is my results:

A test:
```
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon# cp dockerd-1.13.0-dev /usr/bin/dockerd
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon# service docker start
docker start/running, process 3545
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon# docker volume create --name test
test
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon# docker run -d -v test:/data ubuntu:14.04 sleep 100000
92b9f8201bb7ac3a8d4fcfef839c3611cea54afe1ff74a4adf213308a35d0c44
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon# docker version
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:19:35 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7.1
 Git commit:   7b1e733
 Built:        Sat Oct 15 01:24:58 2016
 OS/Arch:      linux/amd64
root@10-11-11-54:~/docker/bundles/1.13.0-dev/binary-daemon#
```
dockerd's commit id is 7b1e733 the same as this pr's commit id.


In addition, in the side of base code in store.go:
```
func (s *VolumeStore) CreateWithRef(name, driverName, ref string, opts, labels map[string]string) (volume.Volume, error) {
	name = normaliseVolumeName(name)
	s.locks.Lock(name)
	defer s.locks.Unlock(name)

	var err error

	v, exist := s.getNamed(name)
	if !exist {
		// it means no volume with name exists, we should create one
		v, err = s.create(name, driverName, opts, labels)
		if err != nil {
			return nil, &OpErr{Err: err, Name: name, Op: "create"}
		}

	}
	s.setNamed(v, ref)

	return v, nil
}
```
Since create a container with a volume, dockerd will call `CreateWithRef`. In the code if `v, exist := s.getNamed(name)` returns a volume, it means the volume exists, then add the container reference and return the existing volume. Then it will not return `errNameConflict `.

I think I should explain more about this PR. please let me know if there is some place still confusing. 🐋

ssuvasanth commented:
@allencloud It looks perfect to me.
PTAL Maintainers.

thaJeztah commented:
/cc @cpuguy83

cpuguy83 commented:
Closing this in favor of #27675
Thanks!
